First round of development

Dentosal opened the pull request:
Adds the following features:
* README.md with setup and usage instructions
* Protocol buffer definition
* gRPC communication
* mTLS certificate generation and use
* Simple job runner that can spawn jobs, stop them, and query status
* Job operation access limited only to the creating client
* CLI
* Some preliminary tests

Not in-scope yet:
* Resource limits and isolation
* Testing  more complex usage patterns and edge cases

Dentosal committed:
Add usage and installation instructions to README

Dentosal committed:
Add a simple git hook manager script

Dentosal committed:
Add certificate generation script and config

Dentosal committed:
Add initial .gitignore

Dentosal committed:
Add project structure and protobufs

Dentosal committed:
Add project skeleton: Tonic, gRPC, mTLS and a CLI

Dentosal committed:
Flesh out the server: endpoints, authorization

Dentosal committed:
Add some end-to-end tests

Dentosal committed:
Run clippy for tests as well

Dentosal committed:
Refactor server service implmentation to own module

Dentosal committed:
Allow configuring server mTLS certificates

Dentosal committed:
Handle receiver hang-up in OutputStream

Dentosal committed:
Improve OutputStream docs and state checking

Dentosal committed:
Improve server code documentation

Dentosal committed:
Improve client library documentation

Dentosal committed:
Add a note about client library docs to README

Dentosal committed:
Improve client output streaming error handling

Dentosal committed:
Refactor OutputStream setup to be generic over input stream

knisbet commented on client/src/bin/cli.rs:
```diff
@@ -0,0 +1,99 @@
+        Operation::Status { jobid } => {
+            pretty_print_status(client.status(JobId::parse(&jobid)?).await?);
+        }
+        Operation::Wait { jobid } => {
```
I don't believe Wait was a listed endpoint in the design (was maybe a potential improvement at the end). Why the deviation from the design?

Dentosal replied:
You're right, it wasn't. When implementing tests, I noticed that it would be way more convenient if the CLI had a status command. By implementing it in the client library, it could be changed to a proper notification instead of loinng without chaning the API at all. However, if you think I should stick to the original design exactly, I can do that as well.

knisbet replied:
We're not sticklers to the design at Teleport, there will always be new ideas or things that come up during implementation, I just wanted to understand why the change was introduced.

knisbet commented on gen-certs.sh:
```diff
@@ -0,0 +1,86 @@
+
+cd server
+openssl genrsa -out server.key 4096
+# openssl req -new -key server.key -subj '/CN=localhost/O=Server/C=US' -out server.csr
```
nitpick: if this isn't needed should probably be cleaned up before merge.

Dentosal replied:
Cleanup in 254da17

knisbet commented on server/src/output_stream.rs:
```diff
@@ -0,0 +1,112 @@
+        let mut index = 0;
+        loop {
+            let h: &OutputHandler = from.borrow();
+            let state = h.state.read().await;
```
I might be missing something since I don't have as deep rust experience as the other reviewers, but won't this hold the read lock through the .await calls below?

Dentosal replied:
Yes, and it will actually block in case of multiple readers. Fixed in 8c5f264. That's sadly a bit complicated to test automatically, but I maybe I should try.

Dentosal replied:
Added a test case for it, in 3236e05 under `common/tests/concurrent.rs`

knisbet commented on server/src/job.rs:
```diff
@@ -0,0 +1,74 @@
+    pub stderr: Arc<OutputHandler>,
+}
+impl Job {
+    pub fn spawn(owner: ClientName, req: JobStartRequest) -> Result<Self, String> {
```
I'm not sure if I just can't find it, but I don't see anything but the status function trying to wait on the child. I suspect this will produce zombies if the client never checks the status of the child process via the API after the child exits.

Dentosal replied:
You're right, this will cause zombie processes. I though that Tokio should be able to handle this by itself, as the documentation says

> The tokio runtime will, on a best-effort basis, attempt to reap and clean up any process which it has spawned. No additional guarantees are made with regards how quickly or how often this procedure will take place.

Looking more into it, this only concerns cases when the  `Child` handle has been dropped. Since this is not the case, the best-effort orphan reaper will not run.

I've fixed this in by 12782b1 managing the child process in a separate task.

knisbet commented on server/src/output_stream.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// Push new data to the history, notifying all waiting processes
+    pub async fn push(&self, data: Vec<u8>) {
```
Not sure push/complete should be public functions as they look internal only, especially since the process can panic if miss-used.

Dentosal replied:
They are not necessarily internal only; if you would like to use `OutputHandler` for something that doesn't implement `AsyncRead` (for instance, stdlib process output), then those would be still useful. The panic conditions are documented in the doc comment. Although they are used here only internally, there is nothing from with them being public (from OOP perspective, anyway). On the other hand, it might be a bit cleaner to keep them private until they are used outside the class, since the whole `server` is a binary crate, and doesn't provide an code-level API anyway.

Dentosal committed:
Cleanup cert generation script

Dentosal committed:
Fix that output streaming would hold an RwLock too long

This caused multiple readers of a single job to block.

Dentosal committed:
Wait for every child process

Dentosal committed:
Add tests for concurrent output and permissions

xacrimon commented on client/src/bin/cli.rs:
```diff
@@ -0,0 +1,99 @@
+
+            while let Some(event) = rx.recv().await {
+                match event? {
+                    (OutputStream::Stdout, data) => out.write_all(&data).unwrap(),
```
This can fail spontaneously due to OS stuff iirc. Should at least have proper error handling and reporting. Unwraps are only to be used when you can prove with a pen and piece of paper and some logic why something cannot panic.

Dentosal replied:
> This can fail spontaneously due to OS stuff iirc.

Yes, it can fail. However, it isn't much use to report errors after that. Where would the report go? To the stdout we cannot write to? I agree that `expect` would still be more readable, and will change to that.

Dentosal replied:
Fixed in 919732f

xacrimon commented on client/src/bin/cli.rs:
```diff
@@ -0,0 +1,99 @@
+    Ok(())
+}
+
+fn pretty_print_status(status: JobStatus) {
```
To follow Rust idioms and guidelines this should be in the form of an implementation of `std::fmt::Display` on `JobStatus`.

Dentosal replied:
Fixed in 93bd033.

xacrimon commented on client/src/lib.rs:
```diff
@@ -0,0 +1,143 @@
+}
+impl Client {
+    /// Connect to a job server
+    pub async fn connect(target: String, tls: TlsConfig) -> DResult<Self> {
```
This should take &str if possible.

Dentosal replied:
The internal client requires owned `String` anyway, so I'm going to use `Into<String>` to accept both. Fixed in 6304652.

xacrimon commented on client/src/lib.rs:
```diff
@@ -0,0 +1,143 @@
+        Ok(response
+            .into_inner()
+            .try_into()
+            .expect("Server returned invalid JobId"))
```
Lacks proper error handling.

Dentosal replied:
This is an unrecoverable critical error. The panic should be acceptable here, unless we anticipate that the communication format will be changing in the future. But then again, we do not have any other version negotiation or forward-compatibility either.

Dentosal replied:
Fixed in 83f768f: Added a better error message, which is returned to the client.

xacrimon commented on client/src/lib.rs:
```diff
@@ -0,0 +1,143 @@
+
+    /// Cancels a job by sending `SIGKILL`.
+    /// This is done asynchronously, and can return before the process has terminated.
+    /// If you must wait until the job has stopped, it can do so by calling `wait`.
```
Calling sigkill on a stopped job makes no sense.

Dentosal replied:
Clarified in d5b1e53.

Dentosal committed:
Implement Display for JobStatus

Dentosal committed:
Use expect instead of unwrap in for output errors

Dentosal committed:
Accept &str target for Client::new

Dentosal committed:
Clarify behavior of stop in the client code

Dentosal committed:
Report client errors nicely with thiserror

knisbet commented on server/src/output_stream.rs:
```diff
@@ -0,0 +1,121 @@
+                break;
+            } else {
+                // Wait until more output is available
+                h.notify.notified().await;
```
What happens if data is pushed to the history between dropping the read lock and registering the waiting task in notify... and then there isn't new data / a new notification for awhile?

Dentosal replied:
That causes a delay until more output is pushed. In worst case the client registers after the job has completed, causing the client to wait indefinitely.

This is clearly a bug. Fixed in f44a564.

Unfortunately, [`tokio::sync`](https://docs.rs/tokio/1.7.0/tokio/sync/index.html) seems to offer no ready solutions to this. Instread, I rolled my own synchronization method, internally using an atomic generation number, and a `oneshot`-channel based update-notification registration.

Dentosal committed:
Fix a race condition in output streaming

If new data was pushed after the reader drops
the read lock but before it actually starts
waiting for notifications, the client will miss
that update and instead waiting for the next one,
which could take really long or even forever if
the missed update was process completion.

knisbet commented on server/src/output_stream/handler.rs:
```diff
@@ -1,35 +1,54 @@
 use std::sync::Arc;
 
 use tokio::sync::mpsc::Sender;
-use tokio::sync::{Notify, RwLock};
```
I think notify would've worked for you, you just need to split up when you call notified() and when you call await, as per https://docs.rs/tokio/1.7.0/tokio/sync/struct.Notify.html#method.notify_waiters

Dentosal replied:
Right, reading the documentation. Should have done that. 

New attempt here, this time only moving the creation around 52c2043

Dentosal committed:
Revert overcomplicated fix to a race condition

This reverts commit f44a56408229c273d5b59c42f3a5ec245b7223af.

Dentosal committed:
Fix a race condition in output streaming (attempt 2)

If new data was pushed after the reader drops
the read lock but before it actually starts
waiting for notifications, the client will miss
that update and instead waiting for the next one,
which could take really long or even forever if
the missed update was process completion.
